test/common/thread_local/thread_local_impl_test.cc

```diff
@@ -82,6 +82,32 @@ TEST_F(ThreadLocalInstanceImplTest, All) {
   tls_.shutdownThread();
 }
 
+// Validate ThreadLocal::runOnAllThreads behavior with all_thread_complete call back.
+TEST_F(ThreadLocalInstanceImplTest, RunOnAllThreads) {
+
+  SlotPtr tlsptr = tls_.allocateSlot();
+
+  EXPECT_CALL(thread_dispatcher_, post(_));
+  EXPECT_CALL(main_dispatcher_, post(_));
+
+  // Ensure that the thread local call back and all_thread_complete call back are called.
+  std::shared_ptr<std::atomic<bool>> all_threads_complete =
+      std::make_shared<std::atomic<bool>>(false);
+  std::shared_ptr<std::atomic<uint64_t>> thread_local_calls =
+      std::make_shared<std::atomic<uint64_t>>(0);
+
+  tlsptr->runOnAllThreads([thread_local_calls]() -> void { ++*thread_local_calls; },
+                          [all_threads_complete, thread_local_calls]() -> void {
+                            EXPECT_EQ(*thread_local_calls, 1);
+                            *all_threads_complete = true;
+                          });
+
+  EXPECT_TRUE(*all_threads_complete);
```

Comment:
this looks racy.  runOnAllThreads returns instantly, calling the 'done' callback when it is done.  The pattern I expect here is for you to establish a condition variable in your test, which you wait on here, and signal that condvar on the done callback.

You can also spin-loop on all_threads_complete since it is an atomic, but IMO it's better to use a condvar and make all_threads_complete a simple bool (not atomic, not shared).

Reply:
I have changed the test to use condvar.